### PR TITLE
Remove official ndebug variable support

### PR DESCRIPTION
**Pull request by Aatch**
[Rendered View](https://github.com/Aatch/rust-rfcs/blob/remove-ndebug/text/0000-remove-ndebug.md)

### CODE REVIEW

**Comment by SimonSapin** [comment]
It can be useful to run tests with both `debug_assert!` and optimizations enabled (if running tests take a lot of time).

**Comment by huonw** [comment]
@SimonSapin I think the suggestion is to allow `-C debug-assertions=true` to override the `opt-level`-based default settings.

**Comment by SimonSapin** [comment]
Ok, that sounds fine.

**Comment by nikomatsakis** [comment]
Yes, this makes sense. `ndebug` in my head has kind of been what you are actually proposing -- that is, a way to enable/disable expensive assertions.

**Comment by nikomatsakis** [comment]
Thinking a bit more, I think it would be useful to have the ability to narrow down the list of modules for which debug-assertions are enabled, as well. I've wanted this in C++ from time to time (and achieved it, by playing with `#undef NDEBUG` and so on). I'd start by just having this controlled via the command-line, but eventually we may want in-code scoped annotations, much like what was proposed for integer overflow initially. My opinion on that, though, is that I'd prefer to take it slow and start simple; I feel we don't have enough experience to judge just how much control is really desired, and I want to avoid over-designing things. To that end, even this proposal of scoping by module is probably overboard: I think all I mean is that I'd like to leave room for specifying a list of modules or other information in the future (categories of debug-assertions and so on).

**Comment by brson** [comment]
In favor of the general idea.

**Comment by brson** [comment]
For context `ndebug` started out as an in-tree thing and sorta seeped into the outside world.

**Comment by sfackler** [comment]
The `debug!` logging macro also disables itself if `ndebug` is set at the moment, but my restructuring of the log crate does away with that in favor of a more configurable system: https://github.com/sfackler/log-ng/blob/master/src/macros.rs

File: text/0000-remove-ndebug.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+time of this writing integer overflow checking is will also be controlled by this variable. Since
+the optimisation setting does not influence `ndebug`, this means that code that the user expects to
+be optimised will still contain the overflow checking logic. Similarly, `debug_assert!` invocations
+are not removed, contrary to what intuition should expect. Enabling optimisations should been seen
+as a request to make the user's code faster, removing `debug_assert!` and other checks seems like
+a natural consequence.
+
+# Detailed design
+
+The `debug_assertions` variable, the replacement for the `ndebug` variable, will be compiler
```
Comment (alexcrichton) [inline text/0000-remove-ndebug.md]: Just clarifying, but by variable do you mean `cfg` directive? For example, I would be able to do:

```rust
if cfg!(debug_assertions) { /* super-expensive code */ }
```

File: text/0000-remove-ndebug.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Remove official support for the `ndebug` config variable, replace the current usage of it with a
+more appropriate 'debug_assertions` compiler-provided config variable.
```
Comment (alexcrichton) [inline text/0000-remove-ndebug.md]: I think markdown doesn't necessarily like this stray tick

**Comment by brson** [comment]
I have a minor concern that `-C debug-assertions` might not be the right place for this command line flag - it doesn't really affect code generation, at least in the current codebase (also `--cfg debug_assertions` has the same affect.

What other things might `debug_assertions` control? Does it affect integer overflow?

**Comment by huonw** [comment]
It seems like the flag could be more than just a boolean, but rather take a list of what to enable to allow fine-grained control, e.g. `none`, `overflow-checks`, `debug_cfg,overflow-checks`, `all`. (Where `-C debug-assertions=debug_cfg` acts like `--cfg debug`.)

**Comment by nikomatsakis** [comment]
Just re-read the RFC again. I like it. I definitely like this part of it:

- `-O0` enables all optional debugging stuff
- `-O1` and higher disables all optional debugging stuff, but you can opt back into it

The only question then is how fine-grained to make the "opting back in". That is, do we want to distinguish debug-assertions from integer-overflow etc? I think we certainly want a simple way to say "put back in *all* the optional debugging stuff". Having more control also seems good but I'd be happy for that to come later too.

**Comment by huonw** [comment]
Sorry for bikeshedding... if we want this to apply to more than just `debug_assert` do we want to use a name other than `-C debug-assertions`?

**Comment by nikomatsakis** [comment]
@huonw I don't have a strong opinion. I'm going to merge the RFC in any case, I think, and we can bikeshed independently and update the RFC/impl as we go. I think the important part is that we plan on integrating this more deeply in the compiler and that we plan on having the defaults be connected to the optimization level.

**Comment by nikomatsakis** [comment]
The core team has decided to merge this RFC. There was no significant opposition, seems like a popular proposal all around. Some [minor] questions as to the precise command-line flags remain to be resolved \[[1], [2], [3]\], but the high-level idea of integrating ndebug more deeply into the compiler and tying it to the optimization flags is uncontroversial.

[1]: https://github.com/rust-lang/rfcs/pull/563#issuecomment-72549694
[2]: https://github.com/rust-lang/rfcs/pull/563#issuecomment-72550619
[3]: https://github.com/rust-lang/rfcs/pull/563#issuecomment-74762795

**Comment by nikomatsakis** [comment]
Tracking issue: https://github.com/rust-lang/rust/issues/22492

**Pull request merged**